pi-top/pi-top-4-Miniscreen
# Handle project exit conditions

| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/OS-1273) |


#### Main changes
Handle project exit conditions:
- For now, the supported options that can be set in a configuration file using the `exit_condition` entry (eg: [here](https://github.com/pi-top/pi-top-4-Miniscreen/blob/fba51153e5e58df8bd938f0165802cb29e033da3/pt_miniscreen/demo_projects/pi_top_4/pong_game/project.cfg#L4)):

1. `POWER_BUTTON_PRESS`: press the pi-top[4] power button. This is now the default exit condition.
2. `HOLD_X`: press the 'X' (Cancel) button of the miniscreen for at least 3 seconds.
3. In the case an invalid or empty string is provided, the project will run until the provided command naturally stops.

- Added a new `Project` class that's in charge of running/waiting/stopping projects. This class is instantiated by a `ProjectPage` whenever the user decides to start a project.

- Updated ProjectPage text when project is started/running/stopped.

- Added exit conditions to pi-top[4] demos.

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
- Install artifacts from https://github.com/pi-top/pi-top-Python-SDK/pull/589 and https://github.com/pi-top/pi-topd/pull/175
- Install artifacts from this PR
- Restart `pt-miniscreen` service with `sudo systemctl restart pt-miniscreen`
- Once the app restarts, go to the Projects menu, access to the 'pi-top[4] Demos' page. 
- The pong game now has a 'power button press' exit condition, so once it starts, press the power button and the project should close.
- The 'hello world' demo now has a 'hold x' exit condition, once it starts hold the X button for more than 3 seconds and the project should close.

#### Tag anyone who definitely needs to review or help
-
